### PR TITLE
feat(cli): new list-registries sub-command and improved error messages

### PR DESCRIPTION
**Pull request by afiune**
This PR contains two features.

## `list-registries` sub-command
A new `list-registries` sub-command that list all Container Registries configured in the Lacework Account.
```
$ lacework vulnerability container list-registries
              CONTAINER REGISTRIES
------------------------------------------------
  gcr.io
  index.docker.io
  1234567890123.dkr.ecr.us-east-2.amazonaws.com
```

### Example: No configured Container Registries
```
$ lacework vulnerability container list-registries
There are no Container Registries configured in your account.

Get started by integrating your Container Registry using the command:

    $ lacework integration create

If you prefer to configure the integration via the WebUI. Log in to your account at:

    https://my-account.lacework.net

Then navigate to Settings > Integrations > Container Registry.
```

## Improved Container Scan Error Messages
We are improving the container scan error messages for a better, more user-friendly messages that are actionable for the user to resolve the issue at hand.

### Error Case: User provides registry not configured
```
$ lacework vulnerability container scan my-registry.example.com my-image latest

ERROR container registry 'my-registry.example.com' not found

Your account has the following container registries configured:

    > gcr.io
    > index.docker.io
    > 1234567890123.dkr.ecr.us-east-2.amazonaws.com

To integrate a new container registry use the command:

    $ lacework integration create
```

### Error Case: User provides container image not found
```
$ lacework vulnerability container scan gcr.io my-image latest

ERROR container image 'my-image:latest' not found in registry 'gcr.io'.

This is most likely a problem with the container registry integration configured
in your account, verify that it was deployed with the right privileges and that
the repository belongs to the provided registry.

To view all container registries configured in your account use the command:

    $ lacework vuln container list-registries
```

### CODE REVIEW

File: api/integrations_ctr_reg.go [hunk 1]
```diff
@@ -124,6 +124,12 @@ func (svc *IntegrationsService) UpdateContainerRegistry(integration ContainerReg
 	return
 }
 
+// ListContainerRegistries lists the CONT_VULN_CFG external integrations available on the Lacework Server
+func (svc *IntegrationsService) ListContainerRegistries() (response ContainerRegIntResponse, err error) {
```
Comment (dmurray-lacework) [inline api/integrations_ctr_reg.go:128]: Suggestion for renaming this function as it doesn't return just the registries as a list of strings like `getContainerRegistries()` does
```suggestion
// ListContainerRegistryIntegrations lists the CONT_VULN_CFG external integrations available on the Lacework Server
func (svc *IntegrationsService) ListContainerRegistryIntegrations() (response ContainerRegIntResponse, err error) {
```

File: cli/cmd/vuln_container.go [hunk 2]
```diff
@@ -819,3 +804,93 @@ func filterSeverity(severity string, threshold string) bool {
...
+    https://%s.lacework.net
+
+Then navigate to Settings > Integrations > Container Registry.
+`
+			return errors.New(fmt.Sprintf(msg, cli.Account))
+		}
+
+		msg := `container registry '%s' not found
+
+Your account has the following container registries configured:
```
Comment (dmurray-lacework) [inline cli/cmd/vuln_container.go:866]: Really like this!

File: cli/cmd/vuln_container.go [hunk 3]
```diff
@@ -754,3 +804,93 @@ func filterSeverity(severity string, threshold string) bool {
...
+	for _, i := range regsIntegrations.Data {
+		// avoid adding empty registries coming from the new local_scanner
+		if i.Data.RegistryDomain == "" {
+			continue
+		}
+
+		// avoid adding duplicate registries
+		if array.ContainsStr(registries, i.Data.RegistryDomain) {
+			continue
+		}
```
Comment (dmurray-lacework) [inline cli/cmd/vuln_container.go:826]: Could reduce this?

`if i.Data.RegistryDomain == "" || array.ContainsStr(registries, i.Data.RegistryDomain){`

File: cli/cmd/vuln_container.go [hunk 4]
```diff
@@ -754,3 +804,93 @@ func filterSeverity(severity string, threshold string) bool {
...
+
+	if strings.Contains(
+		err.Error(),
+		"Could not successfully send scan request to available integrations for given repo and label",
+	) {
+
+		msg := `container image '%s:%s' not found in registry '%s'.
+
+This is most likely a problem with the container registry integration configured
+in your account, verify that it was deployed with the right priviledes and that
```
Comment (dmurray-lacework) [inline cli/cmd/vuln_container.go:885]: typo

```suggestion
in your account, verify that it was deployed with the right privileges and that
```

Comment (dmurray-lacework) [inline cli/cmd/vuln_container.go:885]: typo

```suggestion
in your account, verify that it was deployed with the right privileges and that
```

File: cli/cmd/vuln_container.go [hunk 5]
```diff
@@ -754,3 +804,93 @@ func filterSeverity(severity string, threshold string) bool {
...
+		registries = append(registries, i.Data.RegistryDomain)
+	}
+
+	return registries, nil
+}
+
+// Creates a user-friendly error message
+func userFriendlyErrorForOnDemandCtrVulnScan(err error, registry, repo, tag string) error {
+	if strings.Contains(err.Error(),
+		"Could not find integraion matching the registry provided",
```
Comment (dmurray-lacework) [inline cli/cmd/vuln_container.go:837]: typo

```suggestion
		"Could not find integration matching the registry provided",
```

**Review by dmurray-lacework — CHANGES_REQUESTED** [review]
Looks good Salim. A couple of suggestions.

Really like this feature, especially the new error messages!

File: cli/cmd/vuln_container.go (hunk 3 above)
Comment (dmurray-lacework) [inline cli/cmd/vuln_container.go:826]: Could reduce this?

```suggestion
		// avoid adding empty registries coming from the new local_scanner and avoid adding duplicate registries
		if i.Data.RegistryDomain == "" || array.ContainsStr(registries, i.Data.RegistryDomain){
			continue
		}
```

Comment (afiune) [reply cli/cmd/vuln_container.go:826]: yeah, I was being explicit with the comments to indicate what are we avoid adding to the list of registries, this works as well 👍🏽

File: cli/cmd/vuln_container.go (hunk 5 above)
Comment (afiune) [reply cli/cmd/vuln_container.go:837]: Unfortunately, this is a typo in the platform, so we have to add the same typo in our code 😭

Comment (dmurray-lacework) [reply cli/cmd/vuln_container.go:832]: :(

File: cli/cmd/vuln_container.go [hunk 6]
```diff
@@ -754,3 +804,88 @@ func filterSeverity(severity string, threshold string) bool {
...
+	if strings.Contains(
+		err.Error(),
+		"Could not successfully send scan request to available integrations for given repo and label",
+	) {
+
+		msg := `container image '%s:%s' not found in registry '%s'.
+
+This is most likely a problem with the container registry integration configured
+in your account, verify that it was deployed with the right privileges and that
+the repository belongs to the provided registry.
```
Comment (afiune) [inline cli/cmd/vuln_container.go:881]: @scottford-lw could you give me feedback on this wording here 👆🏽. Thank you!

File: cli/cmd/vuln_container.go [hunk 7]
```diff
@@ -65,6 +70,50 @@ Arguments:
...
+		Short:   "list all Container Registries configured",
+		Long:    `List all Container Registries configured in your account.`,
+		Args:    cobra.NoArgs,
+		RunE: func(_ *cobra.Command, args []string) error {
+			registries, err := getContainerRegistries()
+			if err != nil {
+				return err
+			}
+			if len(registries) == 0 {
+				msg := `There are no Container Registries configured in your account.
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:87]: @afiune  is there a reason why we capitalize `Container Registries`?

File: cli/cmd/vuln_container.go [hunk 8]
```diff
@@ -65,6 +70,50 @@ Arguments:
...
+
+Get started by integrating your Container Registry using the command:
+
+    $ lacework integration create
+
+Or, if you prefer to do it via the WebUI, log in to your account at:
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:93]:
```suggestion
If you prefer to configure the integration via the WebUI. Log in to your account at:
```

File: cli/cmd/vuln_container.go [hunk 9]
```diff
@@ -45,7 +46,11 @@ NOTE: Scans can take up to 15 minutes to return results.
 Arguments:
   <registry>    container registry where the container image has been published
   <repository>  repository name that contains the container image
-  <tag|digest>  either a tag or an image digest to scan (digest format: sha256:1ee...1d3b)`,
+  <tag|digest>  either a tag or an image digest to scan (digest format: sha256:1ee...1d3b)
+
+To list all Container Registries configured in your account:
+
+    $ lacework vuln container list-registries`,
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:53]: should we be using `vuln` or `vulnerability` here? 🤔

File: cli/cmd/vuln_container.go (hunk 8 above)
Comment (scottford-lw) [reply cli/cmd/vuln_container.go:93]: @afiune I see we use that language in other areas in the CLI. I would like @dmurray-lacework to weigh in here too, but starting a sentence with "Or, ..." is not proper. A comma would typically precede "Do it this way, OR that way."

If we are all in agreement, then we should fix that wording everywhere.

**Review by scottford-lw — CHANGES_REQUESTED** [review]
@afiune a few items to review/discuss.

File: cli/cmd/vuln_container.go (hunk 7 above)
Comment (afiune) [reply cli/cmd/vuln_container.go:87]: no, I will change it

File: cli/cmd/vuln_container.go (hunk 8 above)
Comment (afiune) [reply cli/cmd/vuln_container.go:93]: Thank you! Thank you so much, my friend!! I will update all the locations where we have this sentence:
```
cli/cmd/compliance.go:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/cmd/compliance_aws.go:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/cmd/compliance_azure.go:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/cmd/vuln_container.go:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/cmd/vuln_container.go:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/cmd/vulnerability.go:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/docs/lacework_compliance.md:Or, if you prefer to do it via the WebUI, log in to your account at:
cli/docs/lacework_vulnerability_container.md:Or, if you prefer to do it via the WebUI, log in to your account at:
integration/compliance_test.go:Or, if you prefer to do it via the WebUI, log in to your account at:
```

Comment (afiune) [reply cli/cmd/vuln_container.go]: tenor-70325890

File: cli/cmd/vuln_container.go [hunk 10]
```diff
@@ -65,6 +70,50 @@ Arguments:
...
+		Args:    cobra.NoArgs,
+		RunE: func(_ *cobra.Command, args []string) error {
+			registries, err := getContainerRegistries()
+			if err != nil {
+				return err
+			}
+			if len(registries) == 0 {
+				msg := `There are no container registries configured in your account.
+
+Get started by integrating your Container Registry using the command:
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:89]:
```suggestion
Get started by integrating your container registry using the command:
```

File: cli/cmd/compliance_aws.go [hunk 11]
```diff
@@ -50,7 +50,7 @@ Get started by integrating your AWS accounts to analyze configuration compliance
 
     $ lacework integration create
 
-Or, if you prefer to do it via the WebUI, log in to your account at:
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (scottford-lw) [inline cli/cmd/compliance_aws.go:53]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

File: cli/cmd/compliance.go [hunk 12]
```diff
@@ -74,7 +74,7 @@ Get started by integrating one or more cloud accounts using the command:
 
     $ lacework integration create
 
-Or, if you prefer to do it via the WebUI, log in to your account at:
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (scottford-lw) [inline cli/cmd/compliance.go:77]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

File: cli/cmd/compliance_azure.go [hunk 13]
```diff
@@ -82,7 +82,7 @@ Get started by integrating your Azure Tenants to analyze configuration complianc
 
     $ lacework integration create
 
-Or, if you prefer to do it via the WebUI, log in to your account at:
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (scottford-lw) [inline cli/cmd/compliance_azure.go:85]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

File: cli/cmd/vuln_container.go [hunk 14]
```diff
@@ -65,6 +70,50 @@ Arguments:
...
+
+    $ lacework integration create
+
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:93]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

File: cli/cmd/vuln_container.go [hunk 15]
```diff
@@ -754,3 +804,88 @@ func filterSeverity(severity string, threshold string) bool {
...
+		registries, errReg := getContainerRegistries()
+		if errReg != nil {
+			cli.Log.Debugw("error trying to retrieve configured registries", "error", errReg)
+			return errors.Errorf("container registry '%s' not found", registry)
+		}
+
+		if len(registries) == 0 {
+			msg := `there are no container registries configured in your account.
+
+Get started by integrating your Container Registry using the command:
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:846]:
```suggestion
Get started by integrating your container registry using the command:
```

File: cli/cmd/vuln_container.go [hunk 16]
```diff
@@ -754,3 +804,88 @@ func filterSeverity(severity string, threshold string) bool {
...
+
+    $ lacework integration create
+
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:850]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

File: cli/cmd/vuln_container.go [hunk 17]
```diff
@@ -754,3 +804,88 @@ func filterSeverity(severity string, threshold string) bool {
...
+	if strings.Contains(
+		err.Error(),
+		"Could not successfully send scan request to available integrations for given repo and label",
+	) {
+
+		msg := `container image '%s:%s' not found in registry '%s'.
+
+This is most likely a problem with the container registry integration configured
+in your account, verify that it was deployed with the right privileges and that
+the repository belongs to the provided registry.
```
Comment (scottford-lw) [inline cli/cmd/vuln_container.go:881]:
```suggestion
This error is likely due to a problem with the container registry integration 
configured in your account. Verify that the integration was configured with 
Lacework using the correct permissions, and that the repository belongs
to the provided registry.
```

**Review by scottford-lw — CHANGES_REQUESTED** [review]
@afiune I think I got it all this time. 👍

**Comment by afiune** [comment]
@scottford-lw Done!!!!

tenor-221855428

File: cli/cmd/vulnerability.go [hunk 18]
```diff
@@ -99,7 +99,7 @@ To create a new integration use the following command:
 
     $ lacework integration create
 
-Or, if you prefer to do it via the WebUI, log in to your account at:
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (afiune) [inline cli/cmd/vulnerability.go:102]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

File: integration/compliance_test.go [hunk 19]
```diff
@@ -37,7 +37,7 @@ Get started by integrating one or more cloud accounts using the command:
 
     $ lacework integration create
 
-Or, if you prefer to do it via the WebUI, log in to your account at:
+If you prefer to configure the integration via the WebUI. Log in to your account at:
```
Comment (afiune) [inline integration/compliance_test.go:40]:
```suggestion
If you prefer to configure the integration via the WebUI, log in to your account at:
```

**Review by dmurray-lacework — APPROVED** [review]

**Review by scottford-lw — APPROVED** [review]

**Pull request merged**